src/async_fuse/memfs/kv_engine/etcd_impl.rs

```diff
@@ -1,18 +1,246 @@
+    client: etcd_client::Client,
+}
+
+impl Debug for SessionInner {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("SessionInner")
+            .field("closed", &self.closed)
+            .field("lease_id", &self.lease_id)
+            .field("ttl", &self.ttl)
+            .finish_non_exhaustive()
+    }
+}
+
+impl SessionInner {
+    /// Get the closed flag
+    fn is_closed(&self) -> bool {
+        self.closed.load(Ordering::Acquire)
+    }
+
+    /// Set close flag
+    fn close(&self) {
+        self.closed.store(true, Ordering::Release);
+    }
+
+    /// Reopen close flag
+    fn reopen(&self) {
+        self.closed.store(false, Ordering::Release);
+    }
+
+    /// Keep alive the lease
+    #[allow(clippy::pattern_type_mismatch)] // Raised by `tokio::select`
+    pub async fn keepalive(&self, mut close_rx: mpsc::Receiver<()>, token: CancellationToken) {
+        let mut interval = tokio::time::interval(Duration::from_secs(self.ttl.div_ceil(3)));
+
+        if self.is_closed() {
+            error!("lease keep alive stream closed by is_closed flag");
+            return;
+        }
+
+        // Try to clone a client, if the keeper failed, try to reconnect
+        let mut client = self.client.clone();
+        let lease_id = self.lease_id;
+        let (mut keeper, mut lease_keep_alive_stream) =
+            match client.lease_keep_alive(lease_id).await {
+                Ok((keeper, lease_keep_alive_stream)) => (keeper, lease_keep_alive_stream),
+                Err(e) => {
+                    error!("failed to keep alive lease, error={e:?}");
+
+                    return;
+                }
+            };
+
+        // Set the is_closed flag
+        self.reopen();
+
+        // Used to revoke client
+        let mut client = self.client.clone();
+        // Start to keep alive the lease
+        loop {
+            if self.is_closed() {
+                error!("lease keep alive stream closed by is_closed flag");
+                return;
+            }
+
+            tokio::select! {
+                _ = interval.tick() => {
+                    // Try to send a keep alive request
+                    match keeper.keep_alive().await {
+                        Ok(o) => {
+                            info!("keep alive lease, lease_id={lease_id}, {o:?}");
+                        }
+                        Err(e) => {
+                            error!("failed to keep alive lease, error={e:?}");
+                            break;
+                        }
+                    }
+
+                    // Try to parse
+                    match lease_keep_alive_stream.message().await {
+                        Ok(Some(lease_alive_response)) => {
+                            if lease_alive_response.ttl() == 0 {
+                                error!("lease keep alive stream closed");
+                                self.close();
+                                return;
```

Comment:
Replace `return` with `break`,  all breaking points should always call `self.close()`

Reply:
Ok, so in next loop, current task will check with `is_closed()` mark.